### PR TITLE
Improved resolution of circularly dependent components (#130)

### PR DESCRIPTION
**Pull request by keego**
### What
Improved how circularly dependent components are resolved by adopting webpack's resolution methodology: export empty modules first (as placeholders), then reiterate and actually export each module.

### Why
This allows modules to hold references to their dependencies before they actually get exported :)

### Caveats
as [mentioned in #130](https://github.com/diegohaz/arc/issues/130#issuecomment-328640586) this still does *not* solve the `styled(CircularDependentComponent)` problem, BUT it does allow circularly dependent components to properly reference each other and work properly as long as evaluation doesn't occur until after the exports are fully resolved (i.e not having immediately invoked functions requiring dependent components).

### CODE REVIEW

**Comment by diegohaz** [comment]
Well done! Thank you, @keego.

**Comment by Geczy** [comment]
This commit breaks a few things, thoughts?

image

**Comment by diegohaz** [comment]
@Geczy Just looking at the log, I have no idea.

Was that working with the old index?

**Comment by Geczy** [comment]
Yep works with the old index

**Comment by diegohaz** [comment]
Alright, I'm gonna revert this for now until we figure it out.

**Comment by keego** [comment]
@Geczy @diegohaz looks like merging functions into objects doesn't work, I'll take a stab at a fix

**Comment by keego** [comment]
Okay, I did some more in-depth research of how component imports were resolving in my Vue project.

**My issue:** So this solved an interesting issue where I had

`Page` -depends-on-> `NavHeader` -depends-on->`routes.js`-depends-on->`Page`

where `routes.js` was being resolved with a reference to `Page` but that reference was *being evaluated* (`routes: [{ component: Page }]`) to undefined. The object-merge strategy I was using allowed this to instead be resolved as an empty object `{}` which the actual component object was merged into later.

**Actual functionality of the change**: Turns out this change actually solves a niche issue that allows for object exports to be evaluated *before* being fully resolved as long as their actual properties aren't needed until *after* being fully resolved.

**TLDR**: I think theres a fancy solution to make this work with exported functions as well but I don't think the added complexity would be worth it for this project as I doubt this niche functionality is useful for most cases. If an issue does come up that could benefit from this, then I'll tackle improving this and submitting another PR.
